src/python/pants/backend/build_files/fmt/buildifier/rules.py

```diff
@@ -19,15 +19,30 @@ class BuildifierRequest(FmtBuildFilesRequest):
 
 
 @rule(desc="Format with Buildifier", level=LogLevel.DEBUG)
-async def buildfier_fmt(
+async def buildifier_fmt(
     request: BuildifierRequest.Batch, buildifier: Buildifier, platform: Platform
 ) -> FmtResult:
-    result = await _run_buildifier_fmt(request, buildifier, platform)
-    return result
+    buildifier_tool = await download_external_tool(buildifier.get_request(platform))
+    input_digest = await merge_digests_request_to_digest(
+        MergeDigests((request.snapshot.digest, buildifier_tool.digest))
+    )
+    result = await fallible_to_exec_result_or_raise(
+        **implicitly(
+            Process(
+                argv=[buildifier_tool.exe, "-type=build", *request.files],
+                input_digest=input_digest,
+                output_files=request.files,
+                description=f"Run {Buildifier.options_scope} on {pluralize(len(request.files), 'file')}.",
+                level=LogLevel.DEBUG,
+            )
+        ),
+    )
+    return await FmtResult.create(request, result)
 
 
 # Note - this function is kept separate because it is invoked from update_build_files.py, but
-# not as a rule.
+# not as a rule. This function cannot be called from the `buildifier_fmt` rule because other rules
```

Comment:
If this is true, then `_run_buildifier_fmt` is only used in tests? I think we'll also probably have to figure out another way to run the rules in other formatter backends as well since they follow the same pattern.

Is it possible to invoke the `buildifier_fmt` rule in the tests directly without needing this helper function?

Reply:
Nope, it's used in the sources, actually: https://github.com/pantsbuild/pants/blob/a5129432874181e8fcb3390c11ceedb30b8f6fa1/src/python/pants/core/goals/update_build_files.py#L440

Reply:
I see. Is it possible to make that call invoke the `buildifier_fmt` rule instead of the helper function, or does that not work either?

Reply:
Nope, can't call `buildifier_fmt` rule from the `format_build_file_with_buildifier` rule.

```
  File "/home/username/code/pants/src/python/pants/engine/internals/rule_visitor.py", line 277, in _get_byname_awaitable
    raise parse_error(
pants.engine.internals.selectors.GetParseError: Invalid Get. /home/username/code/pants/src/python/pants/core/goals/update_build_files.py:440: Expected an `**implicitly(..)` application as the only keyword input. Failed for Get() in /home/username/code/pants/src/python/pants/core/goals/update_build_files.py.
```